### DISCUSSION TITLE
Windows 95 Minidriver + "SoftGPU" project.

### DISCUSSION THREAD

**Discussion by stuaxo** [dosemu2/dosemu2#2273] [Off-topic]
Off topic (as Windows 95) but it looks like this interesting Windows 95 video driver
http://www.os2museum.com/wp/windows-9x-video-minidriver-hd/

Has evolved into this much more comprehensive driver + bundle of other drivers (3DFX + sound), for various emulators
https://github.com/JHRobotics/softgpu?tab=readme-ov-file

**Comment by stsp** [comment]
Wow! Really cool.
So that completely nullifies my usual
argument of not supporting win9x: that
we won't have directX anyway.
Looks like directX is not a problem with
this driver.
Of course the best the one can have
of win9x under dosemu2, is by booting
native ms-dos and using kvm mode on
x86 - the rest will require re-implementing
winbolt and win4lin together.
kvm-only solution perhaps just requires GEMMIS.
Not something I'm gonna do (unless people
pay a lot of money for that work :), but at least
now there is a point.

**Reply by stuaxo** [reply]
Now SVARDOS has an EDRDOS kernel, that's probably a natural place for any modern go at Winbolt.

I found this driver having a look at automating an install of Win95 in Qemu, using Ansible, but haven't got too far (well I got it started, and then manually working - it turns out you need CPUFIX for it to work on modern CPUs).

If I can get it fully automated, I'll try at least one other emulator that is supposed to work (maybe DOSBOX), from there I could add DOSEMU and perhaps start torturing the install by having dual boot DRDOS (probably SVARDOS DRDOS).

That's probably where I'll get stuck before motivation runs out, though I guess at least there are stuff like Claude.AI and ChatGPT to talk me through the debug process.

In theory I guess I could run Win.com on a working environment vs a working environment in DRDOS and see what it's checking.

In practice I've never used IDA pro, Ghidra, DOSBOX Xs debug tools so I could get stuck very quickly.

**Reply by Torinde** [reply]
> other emulator that is supposed to work

- https://github.com/JHRobotics/softgpu/issues/22
- https://github.com/JHRobotics/softgpu/issues/13

If you can add VMSVGA/VboxSVGA from VirtualBox, that will make it straightforward to adapt for SoftGPU.

> you need CPUFIX

- Patcher9x from the SoftGPU author integrates CPUFIX and other related functionality. 
- https://github.com/JHRobotics/patcher9x/issues/23

**Comment by stsp** [comment]
> Now SVARDOS has an EDRDOS kernel,

Why is that?
Have they dropped freedos?

> that's probably a natural place for any modern go at Winbolt.

AFAIK no winbolt code landed in dr-dos at the end,
so I am not sure why would the one prefer to do the
new development in asm, rather than in C as in freedos
or fdpp.

> In theory I guess I could run Win.com on a working environment vs a working environment in DRDOS and see what it's checking.
> In practice I've never used IDA pro, Ghidra, DOSBOX Xs debug tools so I could get stuck very quickly.

You mean, you want to try re-implementing winbolt yourself?
Or just winglue?
That might be a bolt statement in either case. :)

**Reply by stuaxo** [reply]
SVARDOS only, as I was looking at who on github seemed to have a recent EDRDOS, I'm not sure if they switched or if it's just an rotation.

It's more one extra project on the set of projects that I rotate around.

I reckon.. .it probably is a possibly a very over bolt statement.. I got some way automating Windows 95 installation in ansible.. maybe I'll get the whole way and then the same for the other OSs though even that first thing took a while..

**Reply by Torinde** [reply]
Do you know where to find WinGlue/WinBolt? Or DR-DOS 7.06/7.07? Or DR DOS "Panther" BETA 1 with multi-user security extension?

**Reply by stsp** [reply]
Even if you get those, what's the use?
You can't get the sources.
I don't think any of the above has any
value w/o sources.

**Reply by stuaxo** [reply]
It looks like Matthias Paul did the work to create those, but nobody is able to get a reply out of them, so they would seem to be to be lost.

**Reply by Torinde** [reply]
Another option for inspiration/proof of concept of running Win9x _not from MS-DOS7_: Win4Lin [was able to boot DOS or Win9x from Linux directory](https://github.com/joncampbell123/dosbox-x/issues/5357#issuecomment-2668018310)

**Reply by stsp** [reply]
AFAIK win4lin re-implemented
entire ring0 part of win95. So it
was completely agnostic to any
DOS within. This is what we do
for win31 - under dosemu2 it
can run under any DOS that
doesn't actually support win31
on a real PC.
But decompiling and rewriting
win95 kernel is a too big task.

**Reply by jschwartzenberg** [reply]
There are a lot of comments in the kernel module and kernel patches from NeTraverse that reveal a lot about how it works:
http://web.archive.org/web/20060505035505/http://www.netraverse.com/member/downloads/files/mki-adapter26_1_3_9.patch
http://web.archive.org/web/20060505035500/http://www.netraverse.com/member/downloads/files/Kernel-Win4Lin3-2.6.9.patch

Full list of patches: http://web.archive.org/web/20060329193956/http://www.netraverse.com/member/downloads/kernel_patch.php

I think indeed they indeed had the Windows kernel running in a similar environment to DOSEMU, but they had a lot more insight in its behavior and made sure it all danced around each other carefully :)

**Reply by stsp** [reply]
Hmm, it almost seems as if they
were trying to run windows along
with linux kernel in ring0! Not
quite sure though, but if it is
the case I wonder how reliable
is that.

**Reply by jschwartzenberg** [reply]
It ran quite stable from what I remember. And performance was extreme, much faster than anything else. I think for instance SMP was not supported though.

**Reply by DamjanJovanovic** [reply]
> But decompiling and rewriting win95 kernel is a too big task.

Not really. On any operating system, the amount of system calls is dwarfed by the amount of user-space APIs. For example, when last I checked, Linux had about 500 system calls, FreeBSD had about 150, but the total number of API functions exported from user-space libraries used by any graphical app (eg. libc, libpthread, libX11, libQt, libGL, etc.) is into many thousands.

What's easier, reimplementing thousands of those user-space functions, or reimplementing a few hundred system calls? FreeBSD reimplemented Linux system calls in the FreeBSD kernel, and can successfully run many Linux applications (as unmodified binaries), even large complex ones like web browsers. By comparison, the Wine project has to reimplement all the user-space DLLs, and how complete is it?

And on Windows, most of the magic is in user-space, so if you could reimplement the Windows kernel as a wrapper to the Linux kernel, you could run Windows applications very well.

**Reply by jschwartzenberg** [reply]
@DamjanJovanovic if you could help out with Windows 95 support, that would be great! This has not been attempted before in such a way. but there are indeed tools like Ghidra nowadays that might make this more doable.

I guess it might be useful to also have a working copy of Win4Lin. I believe its executable has references to VWIN32 as well, which I understand is the missing piece of the puzzle. Files and information that could help is here: https://github.com/dosemu2/dosemu2/issues/988

Also see: https://github.com/sudleyplace/DPMIONE/issues/1

If there are more things/ways that could help, just let it know!

**Reply by stsp** [reply]
> What's easier, reimplementing thousands of those user-space functions, or reimplementing a few hundred system calls?

If the user-space functions are
documented while kernel-space
not, then of course the former.
Also the windows kernel does
not only consist of system calls.
You need a particular memory
layout, particular descriptor
layout, lots of in-memory tables
to which user-space peeks directly,
drivers with undocumented ioctls()'s,
and so on. On top of that, you
need to write a scheduler with
unknown properties.

> By comparison, the Wine project has to reimplement all the user-space DLLs, and how complete is it?

If wine goal was to implement
win95-level of compatibility, they'd
win long ago. But Microsoft adds
new stuff faster than they do.

> if you could reimplement the Windows kernel as a wrapper to the Linux kernel, you could run Windows applications very well.

I would estimate that work in
in 10-20K$. And as the result,
you only get the minimal win95
support: no directX, no NDIS.
Every such component has a
rich kernel-space infrastructure,
so re-implementing each, takes
as much.
And once again, even if you do
this all, you'll only barely beat
HX extender, as modern apps
are not going to work on win95
anyway. So I just don't know who
and why will ever work on that.

**Comment by stsp** [comment]
Can anyone clarify what's going on?
I tried to follow the links @Torinde 
posted in other places, and the best
I could find, is this license change:
https://hg.pushbx.org/ecm/edrdos/rev/d4aed5f1b3e6
Is there anything else I am missing?
Or otherwise its completely uninteresting.

**Reply by Torinde** [reply]
I don't get that. What are you looking for?

**Reply by stsp** [reply]
I am looking for that "license
change" due to which you want
to get this winbolt stuff.

**Reply by Torinde** [reply]
Ah, then maybe @ecm-pushbx can help you more than myself.

EDRDOS considers that license sufficient, I assume because it covers "CP/M and **its derivatives**." (some of the [CP/M derivatives](https://web.archive.org/web/20161125115620/http://www.freedos.org/technotes/press/2000-drdos-hist.txt):  CP/M-86, Concurrent DOS, DOS Plus, DR DOS), and isn't restricted to any website anymore, thus they:
- added License you linked
- [removed](https://hg.pushbx.org/ecm/edrdos/rev/7af085747bb6) old Caldera license
  - [removed](https://hg.pushbx.org/ecm/edrdos/rev/8d01d7fe3cc1) non-commercial restriction

**Reply by stsp** [reply]
EDRDOS may consider anything.
But I still don't see a connection between
this and asking Mathias Paul about
releasing winbolt. If he doesn't reply,
I am not surprised. So why should he?

**Reply by ecm-pushbx** [reply]
I believe the relevant link is a forum post from 2007: https://msfn.org/board/topic/109018-windows-98-in-dr-dos/

Particularly this bit:

> Well, I'm not overly optimistic any more, but I still have not completely lost hope, that at some fine day in the future the current owner of the DR-DOS assets may decide that its commercial life is finally over and that it is due to open source the system. This would make it possible to reunificate the different code branches to create the most advanced DR-DOS ever for all its fans. Well, well... ;-)

I documented the license agreement question back in 2023 on my blog: https://pushbx.org/ecm/dokuwiki/blog/pushbx/2023/0821_late_mid_august_work#enhanced_dr-dos

> DR-DOS became free to distribute, modify, and use more than one year ago, with [the new license agreement by DRDOS, Inc. on 2022-07-07](https://pushbx.org/ecm/download/edrdos/license.htm) which refers to "CP/M and its derivatives". While some have argued that DR-DOS is not a derivative of CP/M, I do think it is. During the last week I happened to notice that EDR-DOS [carries copyright statements](https://hg.pushbx.org/ecm/edrdos/file/version-7.01/command/message.asm#l204) that reach back to the year 1976. But of course, in 1976 there was neither 86-DOS nor any of its compatible systems yet. So I take this as an additional hint that DR-DOS is considered a derivative of CP/M.

**Reply by stsp** [reply]
Am I right that "and its derivatives"
appears only in a clarification message
that was supposed to address the site
distribution question?
Then I won't trust that part, and definitely
not extend its scope arbitrary. Even if
"and its derivatives" was intentional
(which is likely not: he was asked to
clarify about web site only), then it can't
be "implicitly" applied to anything.
The fact that something is a derivative,
should be stated directly in some of the
files of that product. Deduction doesn't
help: if it isn't said explicitly, then it is not.

With usual "IANAL" of course.

**Reply by ecm-pushbx** [reply]
> Am I right that "and its derivatives" appears only in a clarification message that was supposed to address the site distribution question?

I think the "Let this paragraph represent a right to ..." effectively supersedes the earlier agreement, which was indeed prompted by the question about the website restriction.

> Then I won't trust that part, and definitely not extend its scope arbitrary. Even if "and its derivatives" was intentional (which is likely not: he was asked to clarify about web site only), then it can't be "implicitly" applied to anything. The fact that something is a derivative, should be stated directly in some of the files of that product. Deduction doesn't help: if it isn't said explicitly, then it is not.
> 
> With usual "IANAL" of course.

I'm no lawyer either. It's your call whether to use EDR-DOS under this license agreement. I think it's fine.

**Reply by stsp** [reply]
> It's your call whether to use EDR-DOS under this license agreement.

Likely not: you are the one who
distribute it. So IMHO its your sole
responsibility, not of your users.
They do not distribute.

And in either case, IMHO this is a
too weak case to approach Mathias Paul.
He is not going to guess why some
vague statement about CPM derivatives
should allow him to open-source winbolt.

**Reply by Torinde** [reply]
> EDR-DOS [carries copyright statements](https://hg.pushbx.org/ecm/edrdos/file/version-7.01/command/message.asm#l204) that reach back to the year 1976.

Good catch.

> Even if "and its derivatives" was intentional

It seems it was intentional - people asked him to clarify if license is restricted to a single website and he could've removed the website ambiguity while keeping the "CP/M technology" wording (which may be interpreted as covering the derivatives), but instead he clarified that part as well with the new wording "CP/M and its derivatives".

In any case I think it'll be useful for somebody to contact Bryan Sparks to ask for binaries and sources that he may have for the various DR products (even unreleased from development phase) - and can ask for further clarification.

**Reply by stsp** [reply]
Yes, he needs to clarify that by
saying "CP/M and its derivatives,
including EDR-DOS and winbolt
projects", or something like this.
Unless there is such an explicit
wording, I wouldn't suggest
contacting Matthias Paul.

And in fact, why not to ask
Bryan Sparks to open-source
winbolt by himself?

**Reply by stuaxo** [reply]
I think I looked into messaging Bryan, but didn't have any luck - but maybe someone knows a better way of contacting him.

**Reply by Torinde** [reply]
> messaging Bryan

Obviously Scott Chapman knows his email (see the license link) - I assume the [people at](mailto:gaby@gaby.de?subject=Unofficial%20CP/M%20Web%20Site) "Unofficial CP/M website" as well.

What I find by searching - Bryan is currently CTO at DRY CANYON HOLDINGS LLC: [1](https://rocketreach.co/bryan-sparks-email_374170092), [2](https://theorg.com/org/swell-cx/org-chart/bryan-sparks)

Notably [here](https://ysa140.net/contact/) it seems they show both email and telephone.

**Reply by stuaxo** [reply]
Cheers, will give this a go.

**Reply by stuaxo** [reply]
Sent a mail, let's see how it goes.
